### ISSUE TITLE
Remote search needs to be refactored

### ISSUE THREAD

**Issue by mouse-reeve** [bookwyrm-social/bookwyrm#2051]
The way I wrote the remote search function is causing all kinds of problems, and needs to be re-thought.

When a book isn't present in the instance database, the instance runs through a list of external data sources (Inventaire, Openlibrary, and other instances) and queries them. This leads to a number of problems:
1. Search is incredibly slow - the queries are performed sequentially with a long timeout, so even a simple search will take the maximum amount of time to execute
2. The connector with the data may not be queried if it isn't high enough in the priority list
3. There are too many other instances to query
4. Instances that are down or otherwise not good data sources (including temporary test instances) are difficult to remove from search

My thought to fix this would be to execute the queries in parallel rather than in series, with a much shorter overall timeout, and return whatever results have come back in time. I don't know what the best way to do this is, or what if any complications doing something like this would have with Django.

The other piece is to add connectors to the admin UI view (it's currently only in the django admin, which is difficult to use and practically only I know about)

**Comment by mouse-reeve** [comment]
I'd love to get feedback from people with some Django experience on what's technically feasible and wise on this! (@cincodenada and @hughrun spring to mind ❤️)

**Comment by hughrun** [comment]
Sorry @mouse-reeve I've been a bit AWOL lately.

This is definitely a pain point (in my view or at least in my usage, probably the biggest pain point) - search is s...l...o...w.

I suggest we might look into using [AIOHTTP](https://docs.aiohttp.org/en/stable/#welcome-to-aiohttp) to run concurrent requests across all connectors. I've not used this in a real project before but I have experimented with it, and it seems to be ideal for this scenario.

I don't really understand how this is set up on the admin side but I assume an instance admin selects which connectors their instance uses?

The way I can see this working would be something like:

1. when a search is submitted, run the search on all connectors that are active on the instance simultaneously using `aiohttp`;
2. include a sensible timeout so a non-performing connector doesn't lock up searches unreasonably;
3. display all results once all sources return a result or an error.

That at least means we run searches in parallel and have a "worst case scenario" time to get an empty result. It's still not idea though - it means most searches will take longer than they need to, because you're waiting for the slowest connection. 

A slightly more aggressive approach could be to display the first result to be returned, keep running the other requests in the background, and provide a button to "search more sources" or similar wording, if the initial result is unsatisfactory. If you press this button then BookWyrm waits if necessary until all connectors have finished and then returns whatever extra results there are.

The above might consistently return poor quality initial results though, if a crappy source is also the fastest server. So an additional option might be to have an admin setting to indicate a source to always wait for and use if it returns anything. e.g. you might think Inventaire is the best source, so prefer to wait for that and display those results at top instead of always returning `crappymetadata.com`'s results. 

I'm not saying I know how to do any of the above, but that's the approach I'd suggest. Happy to noodle around if that's the way you want to go and you want some help.

**Comment by mouse-reeve** [comment]
I encountered that library as well and thought it looked promising, it's gratifying to have that seconded! I don't think I'd want to start with more aggressive approach, since some connectors will return _something_ for just about every search regardless of the ranking of the match (that something often being _The Lord of the Rings_ for Inventaire). 

But I like your idea of having it adjustable to always wait on a preferred connector. I'd have to look more at the library to understand how to use it in practice, but it might be that it wouldn't be too hard to start with the basic approach of having a sensible maximum timeout and then refining it towards a more nuanced approach. It would be hard for the performance to be _worse_ than it currently is, at least!

(And fwiw, no apologies ever necessary about being away -- I hope you and anyone who contributes are only around if it's convenient and fun, and there's no obligation when it's not. It's a such pleasure to have you around though.)

**Comment by hughrun** [comment]
LOL "I couldn't find a match but can I interest you in Lord of the Rings?"

**Comment by cincodenada** [comment]
Hello! I have also been quite busy lately but do intend to continue being around as I can, which should hopefully be more again here in the near future :)

I also agree that running the searches in parallel is probably the biggest win here, and makes a lot of sense. I haven't used aiohttp specifically, but have run into it (and other asyncio-based libraries) before and it seems to be a solid base to build things on. I haven't done much async in Python specifically but I have been doing a lot of async in Javascript lately, and it looks like the basic model in Python is similar in broad strokes at least, and looking at the current remote search, I think it should lend itself to async-ifying relatively easily, and Hugh's outline above aligns with what I was imagining.

I agree that it makes sense to start simple - implementing the basic three-point plan Hugh outlined above (running in parallel with an overall timeout) maps best into the current setup, and should significantly improve the performance, and also forms a good base to start exploring more complex approaches. I also don't think this would require any particular fanciness from Django.

As for the future approaches: I can imagine, for example, a situation where we load results into the page as they come in, instead of waiting to gather all the results we're gonna get and then presenting them at once - that would take more doing, and could involve digging more into Django, depending on how we went about it. I don't know what kind of facilities Django has for that kind of partial loading, but I wouldn't be surprised if they exist.

So in short: another vote from me for starting with the three-point plan Hugh outlined to get things running in parallel to start, and going from there. I may be able to take a run at that this weekend - to get into the details a bit, I'm imagining breaking the body of the [connectors loop](https://github.com/bookwyrm-social/bookwyrm/blob/142616b996ef590499da1ea362c966ccef009282/bookwyrm/connectors/connector_manager.py#L35) into an async function, and then using aiohttp and the [asyncio utility functions](https://docs.python.org/3/library/asyncio-task.html#running-tasks-concurrently) to run the connectors in parallel and cancel them after a given timeout, and then return what we have. Hopefully that makes some sense outside my brain :smile:

**Comment by cincodenada** [comment]
Some further thoughts on approaches: currently there's the option to return the first result we get - is that used, and if so, are the connectors in some specific order, such that it acts as a "preferred connector" ranking? It looks like the asyncio tools are robust enough that we could take all kinds of approaches here. Here's the options I see, some of which are from y'all's comments above. We throw some set of requests out there and then some options are:
 - Wait until they all finish (successful or not) or until a timeout occurs (imo this is Plan A, as it's the simplest to implement and gains much of the benefit)
 - Return the first result we get no matter what, with an option to get more results if those are unsatisfactory (fairly simple frontend implementation, keeping track of pending results on the backend might get tricky)
 - Wait until we have results from one or more of some set of preferred vendor(s), and return anything else we've gotten once we do. Optionally, allow getting more results like the previous item (Similar complexity to above, plus defining preferred vendors)
 - Return results as we get them dynamically (involves a lot more frontend work, probably similar on the backend to above or maybe even simpler)

The other part of this that I think we should hash out is, do we want to limit the number of queries we make right away at some point? Right now that happens naturally, because we only make as many requests as we can before the timeout, but if we're doing them all up front, I think we may not want to search _all_ of the connectors we have, because it's likely that the first few instances we search will come up with something, and it's adding unnecessary load to the searcher and the searchee to be querying them all for every search.

And if you're not querying everyone all the time, do you have a few preferred connectors (e.g. Inventaire) that you always query, and then add in other instances/connectors...randomly? With a preference list? Shuffle it around? Go by some measure of instance size? There's lots of ways we could go there, I think.

**Comment by maxlath** [comment]
@mouse-reeve please let me know next time you encounter an inventaire query that recommends you *Lord of the Rings* for no apparent reason (or similarly weird results), I would be curious to see how that happens on inventaire side ^^

**Comment by maxlath** [comment]
nevermind, I found a case: [any search with `the` or `of`](https://inventaire.io/api/search?types=works&search=the%20mxzjdnfeprlscjjcqkqe&lang=en&limit=10&exact=false) and no better result to offer will suggest you LOTR and other books with a very high popularity that contains those words. Maybe we should filter those very common words out, but there might be cases where that would mean loosing useful results (ex: [It](https://inventaire.io/api/search?types=works&search=it&lang=en&limit=10&exact=false)). Note that if you set the `exact` parameter to `true`, only results containing all the words in the query will be returned. Ex: https://inventaire.io/api/search?types=works&search=the%20mxzjdnfeprlscjjcqkqe&lang=en&exact=true returns nothing

**Comment by mouse-reeve** [comment]
I had the default postgres search configuration that removed filler words and ran into that problem with "There, There" and "It". Good to know about `exact`! But consider returning best-guess results even when the match is slim to be reasonable behavior for a search endpoint :)

**Comment by mouse-reeve** [comment]
> there's the option to return the first result we get - is that used, and if so, are the connectors in some specific order, such that it acts as a "preferred connector" ranking?

This is used exclusively during CSV imports, and currently the connectors are ordered. With a sufficiently efficient remote search, I don't think it would be necessary.
